### ISSUE TITLE
Copy/Paste Causes Dispatcher Warnings and KeyError

### ISSUE THREAD

**Issue by danyeaw** [gaphor/gaphor#585]
**Describe the bug**
Copy and paste is giving me warning and errors.

**To Reproduce**
1. Add a Class to a diagram
2. Click the Undo button:
```
ElementDispatcher WARNING Handler <function Presentation.__init__.<locals>.update at 0x7fc7e0e950d0> is not registered for <gaphor.UML.uml.Class element 768acf82-4bb5-11eb-9208-18cf5efc6bb9>.<association ownedOperation: Operation[0..*] <>-> class_>
```
repeated
The Class is removed from the model.
3. Hit the Redo button and the Class is added back to the model, but is not restored on the diagram.
4. Hit Undo again:
```
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/elementdispatcher.py", line 237, in on_element_change_event
    handler(event)
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/diagram.py", line 239, in _presentation_removed
    self._update_views(removed_items=(event.old_value,))
  File "/home/dan/Projects/gaphor/gaphor/core/modeling/diagram.py", line 351, in _update_views
    v.request_update(dirty_items, dirty_matrix_items, removed_items)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.9/site-packages/gaphas/view/gtkview.py", line 293, in request_update
    self._qtree.remove(item)
  File "/home/dan/Projects/gaphor/.venv/lib/python3.9/site-packages/gaphas/quadtree.py", line 180, in remove
    bounds, data, clipped_bounds = self._ids[item]
KeyError: <gaphor.UML.classes.klass.ClassItem element 768bc3ba-4bb5-11eb-9208-18cf5efc6bb9>
```

**Expected behavior**
The Class is removed from the model for Undo and then restored to the model and diagram upon Redo.

**OS**
- [X] Linux - openSUSE Tumbleweed with GNOME on Wayland
- [ ] MacOS
- [ ] Windows

NOTES:

**Version**
Version of Gaphor: Master

**Additional context**

**Comment by amolenaar** [comment]
Nice catch! It looks like this has to do with changes in the update feature in Gaphas (or the absence thereof).

**Comment by amolenaar** [comment]
The dispatcher warnings are not really warnings. You can set a handler multiple times, and it will only be registered once. Due to circumstances a handler can be unregistered multiple times as well. This is not a problem. The warning is raised, although the desired state is met. I'll lower the level to debug.

**Comment by amolenaar** [comment]
To diagnose Point 3. I did the following:

* Perform step 1 to 3.
* Open the console:
 
```
Gaphor Interactive Python Console
3.8.6 (default, Sep 25 2020, 09:36:53) 
[GCC 10.2.0]
Type "help" for more information.
>>> p=select()[-1].presentation[0]
>>> p
<gaphor.UML.classes.klass.ClassItem element fadc22b3-4c1f-11eb-a938-8fcfae32d12c>
>>> p.matrix
Matrix(1.0, 0.0, 0.0, 1.0, 224.0, 210.0)
>>> p.matrix_i2c
Matrix(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
>>> p.matrix_i2c.set(*p.matrix)
>>> p.update_shapes()
>>> 
```
This will make the shape visible again on the diagram and we can interact with it.

**Comment by amolenaar** [comment]
This issues ties pretty deep into the workings of the undo mechanism. There are some issues i think need to be addressed in order to solve this properly:

1. Undo for Presentation objects are not handled properly. This used to be done in Gaphas, but it is no longer. It kind of works, since the most important relations are associations, but things like .
2. Elements are removed from the model and kinda resurrected (by adding them to `ElementFactory._elements` directly). This should happen via `Element.unlink()` and `ElementFactory.create{_as}()`.
3. In the undo manager, references to objects are kept around. This should be done via id's only (required to make 1. work).
4. The Gaphas state machinery needs to be phased out, then. since it works on object state. Instead we should have our own events, e.g. when items are connected.

Presentation items are not managed by the element factory. I think they should at some point, although I do not see it as a mandatory feature to make undo work again.

What I try to accomplish is that undo/redo of elements is just the same as creating and removing elements. With the only difference that an id is reused. That way we do not have to bother about undo/redo as a different scenario. It's just the same as creating and removing elements.


In Gaphas, the following state is relevant ():
* Line: `orthogonal`, `horizontal` -> replace with attributes?
* Matrix: `invert`, `scale`, `rotate`, `translate` -> some sort of AttributeUpdated event?
* Variable: `value` (also used in position, handles and ports)
* Solver: `add_constraint`, `remove_constraint` -> Connected events?

In Gaphor, `reversible_property` is used in a few places, but it can easily be replaced by an attribute.

**Comment by danyeaw** [comment]
Closed by #588. Nice job @amolenaar!
